### ISSUE TITLE
Got different result with 2 perdict ways(seg task)

### ISSUE THREAD

**Issue by PengPeng-JunJun** [ultralytics/ultralytics#23610]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

I perdict same image by 2 ways. After export onnx, I use yours code perdict and get the same result like by pt file.But when I do it with opencv DNN moduel,the mask pixel all zero.Can you solve my question? This is my opencv code

net = cv2.dnn.readNetFromONNX('E:\\py_Project\\ultralytics\\ultralytics\\yolo\\v8\\segment\\runs\\segment\\train2\\weights\\best.onnx')

image = cv2.imread('E:\\py_Project\\Segmentation\\ENet\\LED2\\20251225115004-554.png')
height, width = image.shape[:2]

blob = cv2.dnn.blobFromImage(image, 1/255.0, (640, 640), (0, 0, 0), swapRB=True, crop=False)

net.setInput(blob)

outs = net.forward()

segmentation_mask = outs[0]  #only one image

if segmentation_mask is None:
    print("Error: Segmentation mask is None.")
else:
    print(f"Segmentation mask shape: {segmentation_mask.shape}")

segmentation_mask = np.argmax(segmentation_mask, axis=0)  

segmentation_mask = (segmentation_mask * 255).astype(np.uint8)

print(f"Resized segmentation mask shape: {segmentation_mask.shape}")
print(f"Segmentation mask dtype: {segmentation_mask.dtype}")

if segmentation_mask.shape != (height, width):
    print("Resizing segmentation mask to match image size...")
    segmentation_mask = cv2.resize(segmentation_mask, (width, height), interpolation=cv2.INTER_LINEAR)

segmented_image = cv2.bitwise_and(image, image, mask=segmentation_mask)

cv2.imshow("Original Image", image)
cv2.imshow("Segmentation Result", segmented_image)

cv2.waitKey(0)
cv2.destroyAllWindows()

**Comment by UltralyticsAssistant** [comment]
👋 Hello @PengPeng-JunJun, thank you for your interest in Ultralytics 🚀! We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

Thanks for sharing the details and your OpenCV DNN snippet 🙏 Since this looks like a 🐛 bug/behavior mismatch between ONNX inference paths (Ultralytics vs OpenCV DNN) for segmentation, could you please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) (MRE) so we can reliably reproduce what you’re seeing? This is especially important for segmentation exports where output formatting/postprocessing can differ across runtimes.

Please include in your MRE:
-Exact `ultralytics` version (`pip show ultralytics`)
-OS, Python, PyTorch, OpenCV versions
-Export command you used (including any args)
-A single test image (or a way to generate a dummy input) and the ONNX model
-The raw `outs` shapes you’re getting from `net.forward()` (e.g. print all output tensors and shapes)
-The exact Ultralytics inference command/code path you’re comparing against

Also, this is an automated response 🤖—an Ultralytics engineer will also assist soon and follow up with more specific guidance once the MRE details are available ✅

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by Y-T-G** [comment]
It uses wrong postprocessing

https://github.com/ultralytics/ultralytics/blob/f65e06bea913aa185c89255cdba5bec6f4d10bad/examples/YOLOv8-Segmentation-ONNXRuntime-Python/main.py#L121

**Comment by PengPeng-JunJun** [comment]
@Y-T-G I want to download yolov8 code again,but conn't find it.Please help me,thank you

**Comment by Y-T-G** [comment]
There's no specific code for YOLOv8. Ultralytics supports multiple models. You can just clone Ultralytics repo.

**Comment by PengPeng-JunJun** [comment]
@Y-T-G My task is seg by onnx on windows system with c++.And I have try opencv dnn and onnxruntime,but failed.Now I only try with openvino.I find the floder,and get some code without seg task.Is anyone do it ?

**Comment by PengPeng-JunJun** [comment]
@Y-T-G I have traind and export onnx by yolov8.So I do not want to give up.So I will try my best do it.

**Comment by Y-T-G** [comment]
You can check this

https://github.com/YYH-GPU/yolov11-yolov8-dnn-Classify-Detect-Obb-Pose-Seg

**Comment by Y-T-G** [comment]
Or this

https://github.com/winxos/yolov8_segment_onnx_in_cpp

**Comment by Y-T-G** [comment]
There are many C++ projects on GitHub 

https://github.com/search?q=yolo+seg+language%3AC%2B%2B+&type=repositories

**Comment by PengPeng-JunJun** [comment]
I got it,thank you!

**Comment by glenn-jocher** [comment]
Glad it’s working now—if you have a moment, please share what change fixed it (especially how you handled the two ONNX outputs for segmentation, i.e. `det` + `proto`, since treating output as a single per-pixel class map will yield all-zero masks in many cases); the expected output structure is described in the [Export mode docs (segmentation output tensors)](https://docs.ultralytics.com/modes/export/).

**Comment by PengPeng-JunJun** [comment]
@Y-T-G Hello, after export onnx with yolov8 seg task train, I see the output by netron, I want to know why the output[0]'s type is 
[1,38,8400].The first is batch_size, it's always 1. I have get some info about the second and third, but can not search them now!

**Comment by Y-T-G** [comment]
You can read this:

https://www.reddit.com/r/Ultralytics/comments/1jwl6se/ultralytics_postprocessing_guide/

**Comment by PengPeng-JunJun** [comment]
Tks! I want to know if you are the Ultralytics offical worker?You are so good at yolo and many problem I can't solve by chatgpt!

**Comment by Y-T-G** [comment]
Yes, I am an Ultralytics employee

**Comment by PengPeng-JunJun** [comment]
@Y-T-G I have do the seg task with custom dataset successful.Include train and inference.Thanks for your help.
But I also have following questions:
1. I labeled the dataset by labelme.So,I must make code  convert json files to yolo format(txt).Is another format I can train by yolo?I have try convert to mask(.png),but when I trainning , the code show error.
2. My labels have circle label.I tried some solutions,only make the circle to polygons can train.Do you have any other solutions?

**Comment by Y-T-G** [comment]
Ultralytics requires TXT files

You can try this:
https://github.com/rooneysh/Labelme2YOLO

**Comment by PengPeng-JunJun** [comment]
I did it like your said.In thos code, I think he also make the circle as a polygon.It seems like a  hexagon.
I get the circle feature also like a polygon.I think it's the postprocess's reason at first.But when I inference with offical cade,the result is same.If so I can't draw the true outline.I want to know the true reason is the dataset convert or the pre model named yolov8n_seg not precise enough?

**Comment by glenn-jocher** [comment]
That “circle looks like a polygon/hexagon” effect is usually not the model but the representation: LabelMe “circle” gets converted into a polygon with a fixed (often small) number of vertices during JSON→YOLO conversion, and YOLO segmentation itself predicts a raster mask at limited internal resolution (proto is much smaller than the input, then upsampled), so contours extracted from `masks.xy` will always be an approximation. If you want a smoother outline, increase the circle-to-polygon sampling points in your converter and try higher-res inference/training (larger `imgsz`) plus `retina_masks=True` to generate masks at native image scale; you can also inspect/use the raw pixel mask (`masks.data`) instead of the contour (`masks.xy`) as shown in the [Isolating Segmentation Objects guide](https://docs.ultralytics.com/guides/isolating-segmentation-objects/).
